File: js/ts/core/lib/calculator.ts
export default class Calculator {
  // ---
  // Basic
  // ---

  /** Add two numbers */
  add(a: number, b: number): number {
    return a + b;
  }
  /** Subtract two numbers */
  subtract(a: number, b: number): number {
    return a - b;
  }
  /** Multiply two numbers */
  multiply(a: number, b: number): number {
    return a * b;
  }
  /** Divide two numbers */
  divide(a: number, b: number): number {
    return a / b;
  }

  // ---
  // Aggregations
  // ---

  /** Get the sum of the provided numbers */
  getSum(numbers: number[]): number {
    return numbers.reduce((acc, curr) => acc + curr, 0);
  }
  /** Get the product of the provided numbers */
  getProduct(numbers: number[]): number {
    return numbers.reduce((acc, curr) => acc * curr, 1);
  }
  /** Get the mean of the provided numbers */
  getMean(numbers: number[]): number {
    return numbers?.length ? this.getSum(numbers) / numbers.length : Number.NaN;
  }
  /** Get the standard deviation of the provided numbers */
  getStd(numbers: number[], completeSample = false): number {
    const n = numbers?.length;
    if (!n) {
      return Number.NaN;
    }
    const mean = this.getMean(numbers);
    const sumOfSquaredDiffs = numbers.reduce(
      (acc, curr) => acc + Math.pow(curr - mean, 2),
      0,
    );
    const populationSize = completeSample ? n : n - 1;
    const std = Math.sqrt(sumOfSquaredDiffs / populationSize);
    return std;
  }

  // ---
  // Random
  // ---

  /** Pretend to get a random integer asynchronously */
  async randInt(max: number): Promise<number> {
    await new Promise(r => setTimeout(r, 100));
    return Math.ceil(Math.random() * max);
  }
}

File: py/testing/pytest-notes/src/calculator.py
from typing import List
from functools import reduce
import math
import asyncio
import random


Num = int | float


class Calculator:
    # ---
    # Basic
    # ---

    @staticmethod
    def add(a: Num, b: Num) -> Num:
        """Add two numbers"""
        return a + b

    @staticmethod
    def subtract(a: Num, b: Num) -> Num:
        """Subtract two numbers"""
        return a - b

    @staticmethod
    def multiply(a: Num, b: Num) -> Num:
        """Multiply two numbers"""
        return a * b

    @staticmethod
    def divide(a: Num, b: Num) -> Num:
        """Divide two numbers"""
        if b == 0:
            return math.inf
        return a / b

    # ---
    # Aggregations
    # ---

    @staticmethod
    def get_sum(numbers: List[Num]) -> Num:
        """Get the sum of numbers"""
        return reduce(lambda acc, curr: acc + curr, numbers, 0)

    @staticmethod
    def get_product(numbers: List[Num]) -> Num:
        """Get the product of numbers"""
        return reduce(lambda acc, curr: acc * curr, numbers, 1)

    @staticmethod
    def get_mean(numbers: List[Num]) -> float:
        """Get the mean of numbers"""
        length = len(numbers)
        if not length:
            return math.nan
        return Calculator.get_sum(numbers) / length

    @staticmethod
    def get_std(numbers: List[Num], complete_sample=False) -> float:
        """Get the std of numbers"""
        n = len(numbers)
        if not n:
            return math.nan
        mean = Calculator.get_mean(numbers)
        sum_of_squared_diffs = reduce(lambda acc, curr: acc + math.pow(curr - mean, 2), numbers, 0)
        population_size = n if complete_sample else n - 1
        std = math.sqrt(sum_of_squared_diffs / population_size)
        return std

    # ---
    # Random
    # ---

    @staticmethod
    async def randint(max_value: int = 100) -> int:
        """Pretend to fetch a random number"""
        if not isinstance(max_value, int):
            raise TypeError()
        if not max_value > 0:
            raise ValueError()

        await asyncio.sleep(.1)
        return random.randint(0, max_value)
